DIFFERENTIAL EXPRESSION ANALYSIS OF GENES
Author: Surya Hembrom
Date: 6.12.22
```{r}
library(dplyr)
library(DESeq2)
library(edgeR)
library(reshape)
library(plyr)
library(tidyverse)
library(arsenal)
library(statmod)
library(plotly)
library(ggfortify)
library(caret)
library(tidymodels)
library(e1071)
library(Boruta)
library(randomForest)
library(ggtext)
library(ggplot2)
```
#the dataset is from PE reads submitted in NCBI GEO dataset as GSE119168
#https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE119168

#download the homo sapiens annotation gff file from Ensembl: https://www.ensembl.org/Homo_sapiens/Info/Index
#this data frame contains the number of reads mapping to exomic regions of gene binned at 50 bp 
```{r}
epitrans.counts <-  gzfile("GSE119168_readCounts.tsv.gz", "rt")
epitrans.counts.df <-  read.table(epitrans.counts, header = TRUE, sep= "\t")

```

```{r}
summary(epitrans.counts.df)

```
#remove the binning and combine the read counts for a given gene
#1. modify the rownames of the dataframe with bins size in the rownames
#create unique list of gene names
```{r}

x <-  list(c(gsub(",.*", "", rownames(epitrans.counts.df))))
names(x) <- "genes"
epitrans.counts.df <- cbind(epitrans.counts.df, x)

```
#sum the read counts per sample based on the genes
```{r}
epitrans.counts.bind.df <- ddply(epitrans.counts.df, "genes", numcolwise(sum))

```
#to get the genes used in this sequencing 
```{r}
x.df <- as.data.frame(epitrans.counts.bind.df$genes)
colnames(x.df) <- "GENES" 
write.table(x.df$GENES, "genelist_dataset.txt")

```
#rename the rownames based on genes from genes column 
```{r}
for (i in 1:length(epitrans.counts.bind.df$genes)){
    rownames(epitrans.counts.bind.df)[i] <- epitrans.counts.bind.df$genes[i]
  }
epitrans.counts.bind.df <- epitrans.counts.bind.df[,-1]

```
#to get the per million reads of each sample for all the genes 
```{r}
epitrans.sample.counts <- as.data.frame(colSums(epitrans.counts.bind.df)/1e6)
colnames(epitrans.sample.counts)[1] <- c("Total_Reads_in_millions")
```
# load the gff file 
```{r}
gff <- read.table("Homo_sapiens.GRCh38.108.chr.gff3.V2.txt")
colnames(gff) <- c("chr_nr", "ensembl", "feature", "start", "end", "strand", "gene_descr", "gene_name")

```
# filter the gff dataframe based on genes found in the dataset
```{r}
name.filtered.df  <-  c("feature", "start", "end", "hgnc_genename")
filtered.gff <- data.frame(matrix(nrow = 0, ncol = length(name.filtered.df)))
colnames(filtered.gff) <- name.filtered.df
filtered.gff <- filter(gff, gene_name %in% x.df$GENES)

filter.epitrans.counts <- with(epitrans.counts.bind.df,
epitrans.counts.bind.df[rownames(epitrans.counts.bind.df)
                       %in% filtered.gff$gene_name,])
```
# gene length 
```{r}
all.gene.length.gff <- as.data.frame(filtered.gff$end- filtered.gff$start)
names(all.gene.length.gff) <-  "gene_length"

filtered.gff <- cbind(filtered.gff, all.gene.length.gff)
colnames(filtered.gff)[9] <- "gene_length"

```
# create DESeq object 
```{r}
condition.disease <- read.csv("sorted_condition_samplev2.txt", header = FALSE, row.names = 1, sep= ",")
names(condition.disease) <- "condition"


met.IP.samples <- data.frame(sampleval =c(rep("input",13), rep("IP",13)))
rownames(met.IP.samples) <- rownames(condition.disease)

d.samples.disease <- cbind(met.IP.samples, condition.disease)
# Convert all columns to factor
d.samples.disease <- as.data.frame(unclass(d.samples.disease),stringsAsFactors = TRUE)               
                      
# dds2 
dds <- DESeqDataSetFromMatrix(countData = filter.epitrans.counts,
                              tidy= FALSE,
                              ignoreRank = FALSE,
                              DataFrame(d.samples.disease),
                               ~(sampleval + condition))
View(counts(dds))

```
# estimates the size factors using the "median ratio method" for each library: normalise based on size factors
```{r}
dds <- estimateSizeFactors(dds)
sizeFactors(dds)

```
# normalised counts matrix 
```{r}
normalised_counts.deseq <- as.data.frame(counts(dds, normalized = TRUE))
pdf("normalisedcounts.pdf", width= 8.27, height = 6.5)
boxplot(normalised_counts.deseq, las =2, cex.axis=0.6, ylab = "CPM")
dev.off()

```
# log 
```{r}
log.counts.deseq <- log2(normalised_counts.deseq+1)
pdf("normalised_logcounts.pdf")
boxplot(log.counts.deseq, las =2, cex.axis=0.6, ylab = "CPM (log2)")
dev.off()
```
#FPKM normalisation: from DESeq2 package: does not work as requires GRanges, genomic ranges for each gene annotated

```{r}
## dds2
#design.disease <- factor(c(condition.disease$condition))
#dds2 <- DESeqDataSetFromMatrix(countData = filter.epitrans.counts,
                              #tidy= FALSE,
                              #ignoreRank = FALSE,
                              #DataFrame(design.disease),
                               #~(design.disease))
#norm.fpkm <- fpkm(dds2)

```


#get the gene names for set of genes from read counts
```{r}
gene.filter.epitrans.counts <- data.frame(rownames(filter.epitrans.counts))
names(gene.filter.epitrans.counts) <- "genes"
write.table(gene.filter.epitrans.counts, "gene_filter_epitrans_counts.txt")
```
# order the filtered.gff file
```{r}
filtered2.gff <- filtered.gff[order(filtered.gff$gene_name), ]

filtered3.gff <- filtered2.gff[, c('gene_name', 'gene_length')]
write.csv(filtered3.gff, "filtered3.gff")
```
# shell scripting:  to generate file with unique values with highest gene length for a given gene
# while read line; 
#do c=`grep -Ei -w "$line" filtered4.gff | sort -k2 -r |awk 'NR== 1'`; 
#echo $c >> filtered_final.txt; done < filtered3.genelist.txt
```{r}
names.final <- c("gene_name", "gene_length")

filtered_final <- read.table("filtered_final.txt")
names(filtered_final) <- names.final

final.genelist <- data.frame(matrix(nrow=0, ncol=2))
names(final.genelist) <- names.final
```
# filter out the genes which are present multiple times in the .gff file due to their various forms present 
# as lnc-gene and pseudogene, so taking the gene feature which is more reliable
```{r}
for (i in 1:dim(gene.filter.epitrans.counts)[1]){
  final.genelist[i,] <- filtered3.gff[filtered3.gff$gene_name %in% gene.filter.epitrans.counts$genes[i], ]
}
```
# compare the order of genes in filter.epitrans.counts and final.genelist from the gff file
```{r}
genenames <- data.frame(final.genelist$gene_name)
comparedf(gene.filter.epitrans.counts, genenames)
all.equal(gene.filter.epitrans.counts, genenames)
comparedf(filtered_final, final.genelist) # both the files produced from the R and shell are same
```
# get the TPM (transcripts per million) instead of RPKM or FPKM as RPKM is only for single reads and 
# TPM is more relevant than FPKM as sum of all TPM for all samples is same 
# see https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/ and https://btep.ccr.cancer.gov/question/faq/what-is-the-difference-between-rpkm-fpkm-and-tpm/ for more details)
# the dataset here is paired end sequenced data so TPM is calculated

#1. calculate the RPK (reads per kilobase) for each gene 
```{r}
rpk <- (filter.epitrans.counts*(10^3)/(final.genelist$gene_length))
```
#2. get the per million scaling factor
```{r}
pm.scalingfactor <- as.data.frame(colSums(rpk)/1e6)
names(pm.scalingfactor) <- "per_million_scaling_factor"
```
#3. calculate the TPM 
```{r}
TPM <- (rpk)/(pm.scalingfactor$per_million_scaling_factor)
```
# check if the TPM of all the samples are the same
```{r}
TPM_persample <- as.data.frame(colSums(TPM))
```
# getTMM 
# create a DGEList based on the grouping of disease condition
```{r}
design.disease <- factor(c(condition.disease$condition))
design.inputIP <- factor(c(met.IP.samples$sampleval))
rpk.norm <- DGEList(counts = rpk, group =  design.disease)
keep <- filterByExpr(rpk.norm)
table(keep)
rpk.norm <- rpk.norm[keep,,keep.lib.sizes=FALSE]
```
# calcNorm : normalize the library size based on scaling factors that minimise the 
# log2 fold changes between samples for genes. 
# see edgeR manual for more details https://bioconductor.org/packages/release/bioc/html/edgeR.html

# Trimmed mean of M values (TMM) calculation for each pair of samples:
## is done for normalisation for removing composition biases between libraries:
```{r}
points <- c(0,1,2,15,16,17)
colors <- rep(c("blue", "darkgreen", "red"), 2)
rpk.norm <- calcNormFactors(rpk.norm, method = "TMM")
pdf("LogFC.pdf")
plotMDS(rpk.norm, cex = 0.6, col=colors[design.disease], pch=points[design.disease], labels = rownames(rpk.norm$samples))
legend("topleft", legend=levels(design.disease), pch=points, col=colors, ncol=2)
dev.off()
```
# convert the normalised rpk into counts per million 
```{r}
counts_GeTMM <- as.data.frame(cpm(rpk.norm))
pdf("counts_GeTMM.pdf", height= 7, width = 9)
boxplot(counts_GeTMM, ylim=c(0,4000), las =2, cex.axis=0.6, ylab="counts TMM")
dev.off()
```
# Log geTMM
```{r}
log.counts_GeTMM <- log2(counts_GeTMM+1)
pdf("counts_GeTMMlog.pdf")
boxplot(log.counts_GeTMM, las =2, cex.axis=0.6, ylab="counts TMM(log2)")
dev.off()
```
####################################################
# Differential Gene Expression Analysis: EdgeR 
#1. create the matrix for the disease condition + input/IP 
```{r}
met.IP <- as.factor(c(rep("input", 13), rep("IP", 13)))
design <- model.matrix(~ design.disease + met.IP)
rownames(design) <- rownames(condition.disease)
```


# NB dispersion values estimation
# plot 
```{r}
data.dispersion <- estimateDisp(rpk.norm, design, robust = TRUE)
pdf("BCVplot.pdf")
plotBCV(data.dispersion)
dev.off()
```
# gene expression analysis from edgeR package: 
#glmQLFit: gene wise negative binomial GLM with quasi likelihood tests
```{r}
fit <- glmQLFit(data.dispersion, design, robust=TRUE)
pdf("QLDisp.pdf")
plotQLDisp(fit)
dev.off()
head(fit$coefficients)
```

# the differentially expressed genes test: quasi-likelihood F-tests:ABORTED due to lesser number of genes acquired that were common to the original paper gene list 
```{r}
# contrast between disease condition and input/IP
con <- makeContrasts(design.diseaseomental_tumor - met.IPIP, levels=design)
qlf <- glmQLFTest(fit, contrast = con)

#pdf("qlf.pdf")
#plotMD(qlf)
#dev.off()

summary(decideTests(qlf))
```
# top down-regulated and up-regulated expressed genes in psoriasis sorted by log-fold change and p value cutoff 0.01
```{r}
topTags(qlf,  sort.by = "logFC", p.value = 0.01) # down and up regulated genes
FDR <- p.adjust(qlf$table$PValue, method="BH") # FDR adjustment with benjamini hochberg test
sum(FDR<0.05) # false discovery rate cut off 0.05
#de.qlf <- as.data.frame(decideTests(qlf))
```

#gene expression glm likelihood ratio tests for differential methylation
#glmLRT:likelihood ratio tests for one or more coefficients in the linear model
```{r}
# contrast between disease condition and input/IP
con <- makeContrasts(design.diseaseomental_tumor - met.IPIP, levels=design)

lrt <- glmLRT(fit, contrast = con)

topTags(lrt, sort.by = "logFC", p.value = 0.01)
summary(decideTests(lrt, adjust.method = "BH", p.value = 0.05))
de.lrt <- as.data.frame(decideTests(lrt))
pdf("lrt.pdf")
plotMD(lrt)
dev.off()
```

# WE TAKE THE RESULTS FROM THE glmLRT TEST AS THE GENES AFFECTED BY THE DIFFERENTIAL METHYLATED REGIONS ARE DIFFERENT THAN THAT PRODUCED FROM THE USUAL DIFFERENTIAL EXPRESSED GENES PRODUCED FROM THE glmQLFTest.
# THE FORMER IS BASED ON GLM LIKELIHOOD TESTS DONE FOR DIFFERENTIAL METHYLATION SEEN NEAR CpG LOCI AND THE
# LATTER IS BASED ON GLM F-TESTS MORE SUITABLE FOR USUAL DIFFERENTIAL EXPRESSION OF GENES.  
# AS WE ARE FOCUSSING ON THE GENOMIC/TRANSCRIPTOMIC REGIONS WHICH GET METHYLATED WE LIMIT TO THE DIFFERENTIAL # METHYLATION EXPRESSION RATHER THAN DIFFERENTIAL EXPRESSION.
# for more details see section 4.7.7 Differential methylation analysis at CpG loci of EdgeR package and 4.3.5 Differential expression

# insignificant genes
```{r}
insignificant.genes <- rownames(de.lrt)[which(de.lrt$`1*design.diseaseomental_tumor -1*met.IPIP` == 0)]
```
# down-regulated genes list 
```{r}
downregulated.genes <- rownames(de.lrt)[which(de.lrt$`1*design.diseaseomental_tumor -1*met.IPIP` ==-1)]
```
# upregulated genes list 
```{r}
upregulated.genes <- rownames(de.lrt)[which(de.lrt$`1*design.diseaseomental_tumor -1*met.IPIP` == 1)]
```
# FILTERING
# transpose the normalised TMM counts
```{r}
transpose.counts_GeTMM <- as.data.frame(t(counts_GeTMM))
```
# filter out the insignificant genes from the normalized TMM transposed counts
```{r}
transpose.counts_GeTMM.filt <- transpose.counts_GeTMM[, !(names(transpose.counts_GeTMM) %in% insignificant.genes)]
```
# pca for upregulated and downregulated genes
```{r}
pca <- prcomp(transpose.counts_GeTMM.filt, scale. = TRUE)
summary(pca)
screeplot(pca)
```
# bind the data frames for PCA analysis
```{r}
transpose.counts_GeTMM.filt.dc<- cbind(transpose.counts_GeTMM.filt, condition.disease)
```
#  plot the PCA of up-regulated and down-regulated genes
```{r}
plot.pca <- autoplot(pca, data = transpose.counts_GeTMM.filt.dc, colour = 'condition') + 
  labs(title = "Upregulated and Downregulated genes") + geom_text(aes_string(x = "PC1", y = "PC2", label = as.factor(rownames(transpose.counts_GeTMM.filt.dc))), color = "black", size= 2.5, nudge_x = 0.015, hjust = 0.1, vjust= 0.25) 
ggplotly(plot.pca)
```

# rename the colnames of genes as Random Forest cannot find it  
```{r}
# rename the gene names 
colnames(transpose.counts_GeTMM.filt.dc) <- gsub("\\-", "\\_", colnames(transpose.counts_GeTMM.filt.dc))

```
# Boruta algorithm for feature selection: Aborted as resulted in very less number of genes that matched with the initial set of genes found in the original paper
#adapted from: https://finnstats.com/index.php/2021/05/03/random-forest/
```{r}
#set.seed(11)
#boruta <- Boruta(as.factor(condition) ~ ., data = reduced.gset.dc, doTrace = 2, maxRuns = 3000)
#boruta <- Boruta(as.factor(condition) ~ ., data = transpose.counts_GeTMM.filt.dc, doTrace = 2, maxRuns = 6000)

#print(boruta)
# do tentative fix of the genes not assigned
#tent.boruta <- TentativeRoughFix(boruta)
#print(tent.boruta)
#tent.boruta.attr <- getSelectedAttributes(tent.boruta, withTentative = F)
```
# plot the boruta: Aborted 
```{r}
#pdf("boruta.pdf", width= 35, height = 20)
#plot(boruta, las = 2, cex.axis = 0.5, xlab = NULL)
#dev.off()
```
# plot importance of boruta: Aborted
```{r}
#plotImpHistory(boruta)
```
# Random forest : supervised learning
###Data partition into training and test 
# scaling of data is not applied to the random forest.

```{r}
set.seed(20)
rf.sample <- initial_split(transpose.counts_GeTMM.filt.dc, prop=0.8) # 80-20% 
rf.train <- training(rf.sample)
rf.test <- testing(rf.sample)

```
# OOB predictor try at each split for random forest
```{r}
n.pred.rf <- ncol(transpose.counts_GeTMM.filt.dc)-1 # removing the condition column and with -1 less predictor
```
# random forest on training data set without boruta algorithm feature selection:
```{r}
set.seed(11)
rf1 <- randomForest(as.factor(condition) ~., data = rf.train, 
                    importance= TRUE, mtry= n.pred.rf, ntree= 10000)
rf1 #OOB error rate is 5%
```
# predict the rf model in the test data:
```{r}
yhat.rf1 <- predict(rf1, rf.test, type = 'response')
```
# confusion matrix:
```{r}
confusionMatrix(yhat.rf1, as.factor(rf.test$condition))
```
# importance of features selected:
```{r}
imp.rf1 <- as.data.frame(importance(rf1, type=1, scale = F)) # get the mean decrease in accuracy
```
# sort the dataframe according to the decrease in the mean decrease accuracy:
```{r}
impsort.rf1 <-  imp.rf1 %>% arrange(desc(MeanDecreaseAccuracy))
# filter the genes 
filtg.rf1 <- subset(impsort.rf1, impsort.rf1$MeanDecreaseAccuracy > 0.0) 
write.csv(rownames(filtg.rf1), "FINAL_lrt_RF_noboruta.csv")
```
# plot the importance of the features selected: 
```{r}
pdf("varImp.pdf", width= 35, height = 20)
varImpPlot(rf1, sort = TRUE, cex= 0.75)
dev.off()
```


# BORUTA + TENTATIVE FIX + RANDOM FOREST
# get features for random Forest from Boruta run: ABORTED
```{r}
#boruta_genes <- getNonRejectedFormula(tent.boruta)
```
# random forest on training dataset with boruta algoritm run on full dataset: Aborted
```{r}
#set.seed(22)
#rf2 <- randomForest(boruta_genes,data = rf.train, importance= TRUE, mtry= length(tent.boruta.attr)-1, ntree= 10000) 
```
# the genes considered important by the boruta algorithm + tentative fix put into random forest 
# for estimation of accuracy of model on test data: Aborted
```{r}
#rf2 # OOB error rate is 0% 
#yhat.rf2 <- predict(rf2, rf.test, type = 'response')
#confusionMatrix(yhat.rf2, as.factor(rf.test$condition))
```
# importance of features selected: Aborted
```{r}
#imp.rf2 <- as.data.frame(importance(rf2, type=1, scale = F)) # get the mean decrease in accuracy
```
# sort the dataframe  according to the mean decrease accuracy: Aborted
```{r}
#impsort.rf2 <-  imp.rf2 %>% arrange(desc(MeanDecreaseAccuracy))
```
# plot the importance: Aborted
```{r}
#pdf("rf2_varimp.pdf")
#varImpPlot(rf2, sort = TRUE, cex= 0.75)
#dev.off()
```
# filter the genes which are important based on mean decrease accuracy: Aborted
```{r}
#filtg.rf2 <- subset(impsort.rf2, impsort.rf2$MeanDecreaseAccuracy > 0.0) 
```
# RENAME THE GENES BACK AGAIN: Aborted
```{r}
#rownames(filtg.rf2) <- gsub("\\_", "\\-", rownames(filtg.rf2))
```
# FINAL GENE LIST AFTER BORUTA+ TENTATIVE FIX +RANDOM FOREST: Aborted
```{r}
#filtg.rf2.list <- as.data.frame(rownames(filtg.rf2))
#names(filtg.rf2.list) <- c("genes")
#write.csv(filtg.rf2.list, "lrt_RF_boruta.csv")
```
# HEATMAP 
# rename the genes of transpose.counts_GeTMM.filt.dc
```{r}
names(transpose.counts_GeTMM.filt.dc) <- gsub("\\_", "\\-", names(transpose.counts_GeTMM.filt.dc))

```
# filter out only the genes selected from random forest rf2 to obtain a dataframe
```{r}
rf1.final.df <- transpose.counts_GeTMM.filt.dc[, (names(transpose.counts_GeTMM.filt.dc) %in% rownames(filtg.rf1))]

```
# add the disease condition as a new column 
```{r}
rf1.final.df.dc <- cbind(rf1.final.df, condition.disease)
```
# COMPLETE DATSET HEATMAP #
# arrange the dataframe based on normal fallopian tube samples followed by omental tumor samples
```{r}
samples.arrange <- c('OvCa1850.input', 'OvCa1850.IP', 'OvCa1917.input',
         'OvCa1917.IP', 'OvCa2005.input', 'OvCa2005.IP',
         'OvCa2013.input', 'OvCa2013.IP', 
         'OvCa2053.input', 'OvCa2053.IP', 'OvCa2064.input', 
         'OvCa2064.IP', 'OvCa2072.input', 'OvCa2072.IP', 
         'OvCa2380.IP','OvCa2380.input', 'OvCa2343.IP',
           'OvCa2343.input', 'OvCa2270.IP', 'OvCa2270.input',
           'OvCa2261.IP', 'OvCa2261.input', 'OvCa2221.IP', 
           'OvCa2221.input','OvCa2186.IP', 'OvCa2186.input')

# arrange the dataframe based on control and disease condition
arr.rf1.final.df.dc <- rf1.final.df.dc %>% arrange(factor(rownames(rf1.final.df.dc), 
                   levels = samples.arrange))
```
# ## melt the dataframe
```{r}
arr.rf1.final.df.dc  <- arr.rf1.final.df.dc %>% select(-condition)
melt.arr.rf1.final.df.dc <- melt(arr.rf1.final.df.dc)
melt.arr.rf1.final.df.dc$samples <- rep(row.names(arr.rf1.final.df.dc),length(arr.rf1.final.df.dc))
```
# normal fallopian tube samples: control 
```{r}
nft = c('OvCa1850.input', 'OvCa1850.IP', 'OvCa1917.input',
  'OvCa1917.IP', 'OvCa2005.input', 'OvCa2005.IP',
  'OvCa2013.input', 'OvCa2013.IP', 
  'OvCa2053.input', 'OvCa2053.IP', 'OvCa2064.input', 
  'OvCa2064.IP', 'OvCa2072.input', 'OvCa2072.IP')
myfilter <- ifelse(melt.arr.rf1.final.df.dc$samples %in% nft, "blue", "red")
```
# blue - normal fallopian tube
# red - omental tumor
```{r}
pdf("finalheatmap.pdf", width = 12, height = 35)
ggplot(melt.arr.rf1.final.df.dc, aes(samples, variable)) +
  geom_tile(aes(fill = value), colour = "white") +
  scale_fill_gradient(low="white", high="black")  +
  theme(axis.text.x = element_markdown(angle = 90, size = 8, color = myfilter), 
        axis.text.y = element_text(size=7)) +
  labs(y= "Genes", x="Samples") + theme(legend.position = "none") 
dev.off()
```
# get the genes which are common to the original gene list given in the original paper and the gene list aquired from this analysis
```{r}

epitrans <- read.table("epitrans.csv", header = FALSE,sep=",")
genes_common <- intersect(epitrans$V1,rownames(filtg.rf1))
write.csv(genes_common, "FINAL_COMMON_EPITRANS.csv")
genes_common.df <- transpose.counts_GeTMM.filt.dc[, (names(transpose.counts_GeTMM.filt.dc) %in% genes_common)]
genes_common.df <- cbind(genes_common.df, condition.disease)

samples.arrange <- c('OvCa1850.input', 'OvCa1850.IP', 'OvCa1917.input',
         'OvCa1917.IP', 'OvCa2005.input', 'OvCa2005.IP',
         'OvCa2013.input', 'OvCa2013.IP', 
         'OvCa2053.input', 'OvCa2053.IP', 'OvCa2064.input', 
         'OvCa2064.IP', 'OvCa2072.input', 'OvCa2072.IP', 
         'OvCa2380.IP','OvCa2380.input', 'OvCa2343.IP',
           'OvCa2343.input', 'OvCa2270.IP', 'OvCa2270.input',
           'OvCa2261.IP', 'OvCa2261.input', 'OvCa2221.IP', 
           'OvCa2221.input','OvCa2186.IP', 'OvCa2186.input')

# arrange the dataframe based on control and disease condition
arr.genes_common.df.dc <- genes_common.df %>% arrange(factor(rownames(genes_common.df), 
                   levels = samples.arrange))

arr.genes_common.df.dc  <- arr.genes_common.df.dc %>% select(-condition)
melt.arr.genes_common.df.dc  <- melt(arr.genes_common.df.dc)
melt.arr.genes_common.df.dc$samples <- rep(row.names(arr.genes_common.df.dc),length(arr.genes_common.df.dc))

nft = c('OvCa1850.input', 'OvCa1850.IP', 'OvCa1917.input',
  'OvCa1917.IP', 'OvCa2005.input', 'OvCa2005.IP',
  'OvCa2013.input', 'OvCa2013.IP', 
  'OvCa2053.input', 'OvCa2053.IP', 'OvCa2064.input', 
  'OvCa2064.IP', 'OvCa2072.input', 'OvCa2072.IP')
myfilter <- ifelse(melt.arr.genes_common.df.dc$samples %in% nft, "blue", "red")

pdf("finalheatmap_reduced.pdf", width = 12, height = 14)
ggplot(melt.arr.genes_common.df.dc, aes(samples, variable)) +
  geom_tile(aes(fill = value), colour = "white") +
  scale_fill_gradient(low="white", high="black")  +
  theme(axis.text.x = element_markdown(angle = 90, size = 8, color = myfilter), 
        axis.text.y = element_text(size=7)) +
  labs(y= "Genes", x="Samples") + theme(legend.position = "none") 
dev.off() 

```
# Differential methylation in gene promoter : TO DO
```{r}
# estimate the dispersion
data.pr <- estimateDisp(rpk.norm, design, trend="none")
data.pr$common.dispersion
# fit the dispersion into glm model
fitpr <- glmFit(data.pr, design)
# apply glm likelihood ratio test
lrtpr <- glmLRT(fitpr)
# top methylated promoter 
topTags(lrtpr,  sort.by = "logFC", p.value = 0.01) # down and up regulated genes
# 
summary(decideTests(lrtpr))

```
